flask_project/campaign_manager/views.py

```diff
@@ -998,143 +907,160 @@ def create_campaign():
             map_type=data['map_type']
             )
         created_campaign.create()
-        save_campaign_feature_types(
-            created_campaign,
-            data
-            )
-        save_campaign_geometry(
-            created_campaign,
-            data
-            )
-        save_campaign_insight_functions(
-            created_campaign,
-            data
-            )
-        save_campaign_managers(
-            created_campaign,
-            data
-            )
+        created_campaign.save_feature_types(data)
+        created_campaign.save_geometry(data)
+        created_campaign.save_insight_functions(data)
+        created_campaign.save_managers(data)
         create_thumbnail_image(created_campaign)
         return redirect(
             url_for(
                 'campaign_manager.get_campaign',
                 uuid=data['uuid'])
         )
-    context = dict(
-        oauth_consumer_key=OAUTH_CONSUMER_KEY,
-        oauth_secret=OAUTH_SECRET,
-        map_provider=map_provider()
-    )
-    context['allowed_managers'] = managers
-    context['url'] = '/create'
-    context['action'] = 'create'
+
+
+def get_campaign_type(campaign_obj):
+    campaign_types = {}
+    for feature_type in campaign_obj.feature_types:
+        campaign_types[feature_type.name] = {}
+        campaign_types[feature_type.name]["type"] = feature_type.name
+        campaign_types[feature_type.name]["feature"] = feature_type.feature
+        campaign_types[feature_type.name]["tags"] = {}
+        for x in feature_type.attributes:
+            campaign_types[feature_type.name]['tags'][x.attribute_name] = []
+    return campaign_types
+
+
+def get_campaign_context(campaign_obj):
+    """ Get required context for edit/copy campaign """
+    context = {}
+    context['allowed_managers'] = campaign_obj.get_managers()
+    context['oauth_consumer_key'] = OAUTH_CONSUMER_KEY
+    context['oauth_secret'] = OAUTH_SECRET
+    context['map_provider'] = map_provider()
+    context['action'] = 'edit'
     context['functions'] = get_selected_functions()
-    context['title'] = 'Create Campaign'
+    context['title'] = 'Edit Campaign'
     context['maximum_area_size'] = MAX_AREA_SIZE
-    context['uuid'] = uuid.uuid4().hex
-    context['types'] = featureTypes
-    context['teams'] = teams
-    context['link_to_omk'] = False
+    context['uuid'] = campaign_obj.uuid
+    context['types'] = {}
+    context['campaign_creator'] = campaign_obj.creator.osm_user_id
+    context['link_to_omk'] = campaign_obj.link_to_OpenMapKit
+    try:
+        context['types'] = json.dumps(
+            get_types()).replace('True', 'true').replace('False', 'false')
+    except ValueError:
+        pass
+    return context
+
+
+def get_campaign_form(campaign_obj):
```

Comment:
this function could be a method of the object Campaign.
`campaign_obj.to_form` (or any name you like)
that you can test.